src/scenic/core/vectors.py

```diff
@@ -4,427 +4,659 @@
-
-	def __getitem__(self, index):
-		return self.coordinates[index]
-
-	def __repr__(self):
-		return f'({self.x} @ {self.y})'
-
-	def __eq__(self, other):
-		if isinstance(other, Vector):
-			return other.coordinates == self.coordinates
-		elif isinstance(other, (tuple, list)):
-			return tuple(other) == self.coordinates
-		else:
-			return NotImplemented
-
-	def __hash__(self):
-		return hash(self.coordinates)
-
-	@classmethod
-	def encodeTo(cls, vec, stream):
-		stream.write(struct.pack('<dd', *vec.coordinates))
-
-	@classmethod
-	def decodeFrom(cls, stream):
-		return cls(*struct.unpack('<dd', stream.read(16)))
+    """A 3D vector, whose coordinates can be distributions."""
+    def __init__(self, x, y, z=0):
+        self.coordinates = (x, y, z)
+        super().__init__(self.coordinates)
+
+    @classmethod
+    def fromSpherical(cls, vector):
+        """ Converts from spherical (rho, theta, phi) to cartesian coordinates"""
+        rho, theta, phi = vector
+        x = rho * cos(phi) * sin(theta)
+        y = rho * sin(phi) * sin(theta)
+        z = rho * cos(theta)
+
+        return Vector(x,y,z)
+
+    @property
+    def x(self) -> float:
+        return self.coordinates[0]
+
+    @property
+    def y(self) -> float:
+        return self.coordinates[1]
+
+    @property
+    def z(self) -> float:
+        return self.coordinates[2]
+
+    def toVector(self) -> Vector:
+        return self
+
+    @staticmethod
+    def _canCoerceType(ty):
+        return issubclass(ty, (tuple, list, numpy.ndarray)) or hasattr(ty, 'toVector')
+
+    @staticmethod
+    def _coerce(thing) -> Vector:
+        if isinstance(thing, (tuple, list, numpy.ndarray)):
+            l = len(thing)
+            if not 2 <= l <= 3:
+                raise CoercionFailure('expected 2D/3D vector, got '
+                                      f'{type(thing).__name__} of length {l}')
+            return Vector(*thing)
+        else:
+            return thing.toVector()
+
+    def sampleGiven(self, value):
+        return Vector(*(value[coord] for coord in self.coordinates))
+
+    def evaluateInner(self, context):
+        return Vector(*(valueInContext(coord, context) for coord in self.coordinates))
+
+    @vectorOperator
+    def applyRotation(self, rotation):
+        if not isinstance(rotation, Orientation):
+            return NotImplemented
+        r = rotation.getRotation()
+        return Vector(*r.apply(self.coordinates))
+
+    @vectorOperator
+    def cartesianToSpherical(self):
+        """Returns this vector in spherical coordinates (rho, theta, phi)"""
+        rho = math.hypot(self.x, self.y, self.z)
+        theta = math.atan2(self.y, self.x) - math.pi/2
+        phi = math.atan2(self.z, math.hypot(self.x,self.y))
+        return Vector(rho, theta, phi)
+
+    @zeroPreservingVectorOperator
+    def rotatedBy(self, angleOrOrientation) -> Vector:
+        """Return a vector equal to this one rotated counterclockwise by the given angle."""
```

Comment:
Should update docstring now that this accepts an orientation as well as an angle.